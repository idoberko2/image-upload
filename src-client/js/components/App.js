// external
import React from 'react';
import { CSSTransition } from 'react-transition-group';
import styled from '@emotion/styled';
import { Global } from '@emotion/core';

// utils
import validateFiles from '../utils/validateFiles';
import sendFiles from '../utils/sendFiles';

// components
import globalCss from './common/globalCss';

import Success from './pages/Success';
import Form from './pages/Form';

const uploadFormInitialState = {
    collection: '',
    collectionChanged: false,
    files: null,
    formVisible: true,
    isLoading: false,
    photographer: '',
    successVisible: false,
    submitError: null,
};

class App extends React.Component {
    constructor(props) {
        super(props);

        this.filesRef = null;
        this.state = uploadFormInitialState;

        this.handleCollectionChange = this.handleCollectionChange.bind(this);
        this.handlePhotographerChange = this.handlePhotographerChange.bind(
            this
        );
        this.handleFilesSelection = this.handleFilesSelection.bind(this);
        this.isCollectionValid = this.isCollectionValid.bind(this);
        this.isFilesValid = this.isFilesValid.bind(this);
        this.isValid = this.isValid.bind(this);
        this.handleSubmit = this.handleSubmit.bind(this);
    }

    handleCollectionChange(event) {
        const collection = event.target.value;
        this.setState(prevState => ({
            collection,
            collectionChanged:
                prevState.collectionChanged || Boolean(collection),
        }));
    }

    handlePhotographerChange(event) {
        const photographer = event.target.value;
        this.setState({ photographer });
    }

    handleFilesSelection(event) {
        const { files } = event.target;

        if (!files || files.length === 0) {
            return;
        }

        this.setState({
            files,
        });
    }

    isCollectionValid() {
        return !this.state.collectionChanged || this.state.collection !== '';
    }

    isFilesValid() {
        return (
            !window.FileReader ||
            !window.Blob ||
            (this.state.files && validateFiles(this.state.files))
        );
    }

    isValid() {
        return this.isCollectionValid() && this.isFilesValid();
    }

    handleSubmit() {
        this.setState({ isLoading: true }, () => {
            sendFiles(this.state)
                .then(() => {
                    this.filesRef.value = '';
                    this.setState(
                        Object.assign(
                            {},
                            uploadFormInitialState, // to reset the form
                            { formVisible: false } // to hide the form and show success message
                        )
                    );
                })
                .catch(err =>
                    this.setState({ submitError: err, isLoading: false })
                );
        });
    }

    render() {
        const {
            collection,
            files,
            formVisible,
            isLoading,
            photographer,
            successVisible,
            submitError,
        } = this.state;

        return (
            <>
                <Global styles={globalCss} />
                <Wrapper>
                    <CSSTransition
                        in={formVisible}
                        timeout={{ enter: 500, exit: 300 }}
                        classNames="fade-in-out"
                        unmountOnExit
                        onExited={() => this.setState({ successVisible: true })}
                    >
                        {() => (
                            <Form
                                collection={collection}
                                submitError={submitError}
                                files={files}
                                filesRef={c => (this.filesRef = c)}
                                isCollectionValid={this.isCollectionValid()}
                                isFilesValid={this.isFilesValid()}
                                isSubmitDisabled={!this.isValid() || isLoading}
                                isLoading={isLoading}
                                onCollectionChange={this.handleCollectionChange}
                                onPhotographerChange={
                                    this.handlePhotographerChange
                                }
                                onFilesChange={this.handleFilesSelection}
                                onSubmit={this.handleSubmit}
                                photographer={photographer}
                            />
                        )}
                    </CSSTransition>
                    <CSSTransition
                        in={successVisible}
                        timeout={{ enter: 500, exit: 300 }}
                        classNames="fade-in-out"
                        unmountOnExit
                        onExited={() => this.setState({ formVisible: true })}
                    >
                        {() => (
                            <Success
                                onReset={() =>
                                    this.setState({
                                        successVisible: false,
                                    })
                                }
                            />
                        )}
                    </CSSTransition>
                </Wrapper>
            </>
        );
    }
}

const Wrapper = styled.div`
    width: 100%;
    height: 100%;

    display: flex;
    align-items: center;
    justify-content: center;
`;

export default App;
