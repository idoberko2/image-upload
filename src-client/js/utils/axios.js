// external
import axios from 'axios';

export default axios;
