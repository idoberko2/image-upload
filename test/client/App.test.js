// external
import React from 'react';
import { mount } from 'enzyme';

// axios mock
import axios from 'axios';
import MockAdapter from 'axios-mock-adapter';
const axiosMock = new MockAdapter(axios);

// components
import App from '../../src-client/js/components/App';
import {
    MyInnerForm,
    SubmissionError,
} from '../../src-client/js/components/pages/Form';

const findInputById = (wrapper, id) =>
    wrapper.findWhere(n => n.name() === 'TextInput' && n.prop('id') === id);

describe('App', () => {
    test('submits correctly', done => {
        const wrapper = mount(<App />);
        const collectionInput = findInputById(wrapper, 'collection');
        const galleryNameInput = findInputById(wrapper, 'galleryName');
        const seasonInput = findInputById(wrapper, 'season');
        const photographerInput = findInputById(wrapper, 'photographer');
        const uploaderInput = wrapper.find('input#uploader');
        const uploadTestFile = new File(['test1'], 'test1.jpg', {
            type: 'image/jpeg',
        });

        collectionInput.simulate('change', {
            target: { name: 'collection', value: '  collection name         ' },
        });
        galleryNameInput.simulate('change', {
            target: { name: 'galleryName', value: '     gallery name     ' },
        });
        seasonInput.simulate('change', {
            target: { name: 'season', value: '     2018-2019        ' },
        });
        photographerInput.simulate('change', {
            target: { name: 'photographer', value: ' photographer name ' },
        });
        uploaderInput.simulate('change', {
            target: {
                name: 'files',
                files: [uploadTestFile],
            },
        });

        axiosMock.onPost().replyOnce(200);

        wrapper
            .find(MyInnerForm)
            .find('form')
            .simulate('submit', {
                preventDefault: () => {},
            });

        setImmediate(() => {
            expect(axiosMock.history.post.length).toBe(1);

            const actualBody = {};
            axiosMock.history.post[0].data.forEach((value, key) => {
                actualBody[key] = value;
            });
            expect(actualBody).toMatchSnapshot();
            expect(actualBody.images).toBe(uploadTestFile);
            done();
        });
    });

    test('handles failure correctly', done => {
        const wrapper = mount(<App />);
        const collectionInput = findInputById(wrapper, 'collection');
        const galleryNameInput = findInputById(wrapper, 'galleryName');
        const seasonInput = findInputById(wrapper, 'season');
        const photographerInput = findInputById(wrapper, 'photographer');
        const uploaderInput = wrapper.find('input#uploader');

        collectionInput.simulate('change', {
            target: { name: 'collection', value: 'c' },
        });
        galleryNameInput.simulate('change', {
            target: { name: 'galleryName', value: 'g' },
        });
        seasonInput.simulate('change', {
            target: { name: 'season', value: '2018-2019' },
        });
        photographerInput.simulate('change', {
            target: { name: 'photographer', value: 'photographer name' },
        });
        uploaderInput.simulate('change', {
            target: {
                name: 'files',
                files: [
                    new File(['test1'], 'test1.jpg', {
                        type: 'image/jpeg',
                    }),
                ],
            },
        });

        axiosMock.onPost().replyOnce(500);

        wrapper
            .find(MyInnerForm)
            .find('form')
            .simulate('submit', {
                preventDefault: () => {},
            });

        setImmediate(() => {
            expect(wrapper.find(SubmissionError).text()).toMatchSnapshot();

            done();
        });
    });
});
