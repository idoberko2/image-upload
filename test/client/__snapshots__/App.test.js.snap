// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App matches snapshot 1`] = `
.emotion-36 {
  width: 100%;
  height: 100%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
}

.emotion-34 {
  margin: 0;
  padding: 0;
  width: 20em;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.emotion-10 {
  margin: 0;
  padding: 0;
  padding-top: 2em;
  border-bottom: 1px solid #adadad;
}

.emotion-10:first-of-type {
  padding-top: 0;
}

.emotion-10:last-of-type {
  border-bottom: 0;
}

.emotion-2 {
  font-size: 4em;
  color: var(--main-color);
  margin: 0 0 0.1em;
}

.emotion-0 {
  font-size: 0.4em;
  font-weight: normal;
  margin: 0 0.5em;
}

.emotion-4 {
  opacity: 1;
  border: none;
  border-radius: 0.5em;
  padding: 0.7em 1em;
  width: 100%;
  font-size: 1.3em;
  border: 1px solid #a2a2a2;
  -webkit-transition: opacity 200ms;
  transition: opacity 200ms;
}

.emotion-4:disabled {
  opacity: 0.7;
}

.emotion-6 {
  opacity: 1;
  border: none;
  border-radius: 0.5em;
  padding: 0.7em 1em;
  width: 100%;
  font-size: 1.3em;
  border: 1px solid #a2a2a2;
  -webkit-transition: opacity 200ms;
  transition: opacity 200ms;
  margin-top: 0.5em;
}

.emotion-6:disabled {
  opacity: 0.7;
}

.emotion-8 {
  height: 2em;
  margin-top: 0.5em;
}

.emotion-17 {
  overflow: hidden;
  position: relative;
  width: 100%;
  display: block;
  color: rgb(231,231,231);
  background-color: var(--main-color);
  font-size: 1.3em;
  padding: 0.7em 0;
  border-radius: 0.5em;
  border: 0;
  -webkit-transition: all 200ms;
  transition: all 200ms;
  cursor: pointer;
  text-align: center;
  width: 100%;
  border: 1px solid var(--main-color);
  opacity: 1;
  -webkit-transition: opacity 200ms;
  transition: opacity 200ms;
}

.emotion-17:hover {
  background-color: var(--secondary-color);
}

.emotion-17:active {
  background-color: var(--secondary-color-light);
  color: #333;
}

.emotion-16 {
  display: block;
  font-size: 999px;
  -webkit-filter: alpha(opacity=0);
  filter: alpha(opacity=0);
  min-height: 100%;
  min-width: 100%;
  opacity: 0;
  position: absolute;
  right: 0;
  text-align: right;
  top: 0;
  cursor: pointer;
}

.emotion-28 {
  width: 100%;
}

.emotion-27 {
  color: rgb(231,231,231);
  background-color: var(--main-color);
  font-size: 1.3em;
  padding: 0.7em 0;
  border-radius: 0.5em;
  border: 0;
  -webkit-transition: all 200ms;
  transition: all 200ms;
  cursor: pointer;
  text-align: center;
  width: 100%;
  border: 1px solid var(--main-color);
  opacity: 1;
  -webkit-transition: opacity 200ms;
  transition: opacity 200ms;
  cursor: default;
  opacity: 0.7;
}

.emotion-27:hover {
  background-color: var(--secondary-color);
}

.emotion-27:active {
  background-color: var(--secondary-color-light);
  color: #333;
}

.emotion-27:hover {
  background-color: var(--main-color);
}

.emotion-27:active {
  background-color: var(--main-color);
  color: #333;
}

<div
  class="emotion-36 emotion-37"
>
  <div
    class="emotion-34 emotion-35"
  >
    <div
      class="emotion-10 emotion-11"
    >
      <h1
        class="emotion-2 emotion-3"
      >
        1
        <span
          class="emotion-0 emotion-1"
        >
          ממלאים פרטים
        </span>
      </h1>
      <input
        class="emotion-4 emotion-5"
        id="collection"
        placeholder="שם האלבום"
        type="text"
        value=""
      />
      <input
        class="emotion-6 emotion-5"
        id="photographer"
        placeholder="שם הצלם"
        type="text"
        value=""
      />
      <div
        class="emotion-8 emotion-9"
      />
    </div>
    <div
      class="emotion-10 emotion-11"
    >
      <h1
        class="emotion-2 emotion-3"
      >
        2
        <span
          class="emotion-0 emotion-1"
        >
          בוחרים תמונות
        </span>
      </h1>
      <label
        class="emotion-17 emotion-18"
        for="uploader"
      >
        בחירת תמונות
        <input
          accept="image/*"
          class="emotion-16"
          id="uploader"
          multiple=""
          type="file"
        />
      </label>
      <div
        class="emotion-8 emotion-9"
      />
    </div>
    <div
      class="emotion-10 emotion-11"
    >
      <h1
        class="emotion-2 emotion-3"
      >
        3
        <span
          class="emotion-0 emotion-1"
        >
          מעלים
        </span>
      </h1>
      <div
        class="emotion-28 emotion-29"
      >
        <button
          class="emotion-27"
          data-testid="submit-button"
          disabled=""
        >
          העלאה
        </button>
      </div>
      <div
        class="emotion-8 emotion-9"
      />
    </div>
  </div>
</div>
`;
